### ISSUE TITLE
Add run.analysisScope

### ISSUE THREAD

**Issue by michaelcfanning** [oasis-tcs/sarif-spec#180]
Suggestion from Semmle review of EN05. Also, consider use of 'observation target' to clarify situation when a result physical location diverges from a configured scan scope (such as when a result occurs in a header file when the configured scope of the analysis was an object file that included it).

**Comment by lgolding** [comment]
@michaelcfanning Are you saying that `result.observationTarget` is another candidate (in addition to `result.analysisScope`) for a new name for `result.analysisTarget`? Or are you saying that "observation target" is a distinct concept?

Perhaps it will become clear when I read EN05, which I will now...

**Comment by lgolding** [comment]
@michaelcfanning I've just searched the word "target" in the spec. Changing "analysis target" to "analysis scope" would require significant wordsmithing; this is not a simple find/replace. Here's an  example from the spec where "analysis scope" just doesn't work:

> A `run` object **MAY** contain a property named `architecture` whose value is a string that specifies the hardware architecture at which the analysis scopes are targeted. 

As to "observation target" -- my problem with that is that it introduces a new term "observation" for something that already has a name: "result". I don't like the "target" part of the name either. I like "analysis target" because it's the thing you _point the tool at_. But you don't point anything at the observation/result target. My preferred name for this would be "result location" -- but we already call it that.

Bottom line: of the five open CSD.1 issues, this is the one that seems least valuable to me, and at the same time is the most costly. I'm going to do the other four first.

**Comment by michaelcfanning** [comment]
Sure. Hold tight. I will review existing spec language.

**Comment by michaelcfanning** [comment]
I agree with the basic premise that in many cases the thing pointed to by a scanner represents a scope. I think observation target is less compelling, basically agreeing with your opinion.

Can we look at renaming analysisTarget to analysisScope and through some spec magic permit the use of target elsewhere in the document? Just a thought.

**Comment by lgolding** [comment]
> Can we look at renaming analysisTarget to analysisScope and through some spec magic permit the use of target elsewhere in the document?

Of course anything is possible. But I still don't understand the objection to "scan target". In every case we've discussed (a folder, a file share, a process, a source file, a binary file, a web site...), we are pointing a scan tool at _something_, and "target" is a good word for "the thing I'm pointing at". In _some_ (but not all!) of those cases, "the thing I'm pointing at" could be considered a scope. So why not use the term that _always_ makes sense?

Since you say this request comes from the Semmle review, I'd like to ask Luke to read this thread and clarify why he thinks "scope" is preferable to "target."

**Comment by lgolding** [comment]
But before I loop Luke in I'm going to re-read the email where they made that suggestion, to see if I can better understand why.

**Comment by lgolding** [comment]
Ok, I have an email where you wrote this

> I’ve attached a revised version of the engineering note around populating SARIF for effective result matching. Pavel, thanks very much for the useful review. I acted on nearly everything you raised, with one exception, see below. Some comments on your feedback:
>
> •	I have split ‘scan target’ into two concepts, a ‘scan scope’ (something scanned by a tool) and ‘observation target’ (something with which a scan tool associates a result, which may or may not have been explicitly specified as a scan scope). Larry, we should consider whether result.analysisTarget should be renamed result.analysisScope.

But I don't have Pavel's initial feedback where he explained why he thought it was a good idea.

**Comment by pavgust** [comment]
My feedback about the term 'scan target' was about how the versions of EN05 and EN06 that I was reading used it. Here's what I wrote:

> I struggled with “Scan target”. The definition there is pretty broad [...]. I also think the text (of this and EN06) uses “Scan target” in two different ways, firstly to mean “something scanned by an automatic checking tool” and secondly to mean “something to which an automatic checking tool has attached a result”. I would suggest “Scan scope” for the former and perhaps something like “Observation target” or “Program element” for the latter.

I think the main point of concern was the use of "scan target" for what I think the spec calls a "result location". For example, EN05 said "... annotations applied directly to a scan target (such as a class declaration)" or "It can be helpful to refer to a scan target by its logical name, i.e., a fully-
qualified path that identifies a scan target within a container.", where, in context, I think it did not mean "the set of things that were scanned". I mainly wanted to have a separate term for "something you can attach a result/observation to", and the best suggestions I had at the time were "observation target" and "program element".

I agree with the points made in this thread so far -- "analysis scope" seems like a reasonable name to use in the spec, and observations that a tool might make can reasonably be called "results" for SARIF purposes.

**Comment by lgolding** [comment]
Thanks @pavgust . Ok, I think I have my arms around this now. Regardless of how EN05 or EN06 may have used or abused :-) the terms, I think there are three concepts here. Two of those concepts exist in the spec (and I think they are well named). The third concept is missing and we should add it to the spec:

1. THE MISSING CONCEPT: The set of artifacts that the tool was configured to analyze. It might be a directory full of source code, a file share, a process, a web site, or anything else. Let us call this concept "**analysis scope**". We might capture this in a new run-level property `run.analysisScope`. 

    EXAMPLE:
```"
{                                                  # A run object.
  analysisScope": "src/**/*.c",
  ...
}
```

2. The artifact that the tool was analyzing when it detected the result. The spec refers to this concept as "**analysis target**". It is captured in the existing result-level property `result.analysisTarget`.

    EXAMPLE:
```
{                                                  # A result object.
  "analysisTarget": {
    "uriBaseId": "SRCROOT",
    "uri": "ui/window.c"
  },
  ...
}
```

3. The artifact in which the tool detected the result. The spec refers to this concept as "**result file**". The result file is captured as _part_ of the result-level property `result.location` (which can also include line and column information for file-based results, but it doesn't have to):

    EXAMPLE:
```
{                                                  # A result object.
  "location": {
    "fileLocation": {
      "uriBaseId": "SRCROOT",
      "uri": "include/rectangle.h"
    },
    "region": {
      "startLine": 42
    }
  },
  ...
}

@michaelcfanning

**Comment by michaelcfanning** [comment]
Also consider source code archives, UNC shares, tar/other archives, a NuGet package before resolving this.

**Comment by michaelcfanning** [comment]
See closed item #142, which covers the scenario of specifying directories as analysis scope.

**Comment by lgolding** [comment]
I restored this issue's original name, and marked the issue `resolved-by-design`, to emphasize that we have considered and rejected the request in its original form.

In analyzing that request, above, I pointed out that there was in fact a new concept of "analysis scope", _distinct from "analysis target"_, that SARIF does not represent. The newly file #294, "Add run.scope", captures that concept.
